src/dataModels/JSON.js

```diff
@@ -313,63 +274,84 @@ var JSON = DataModel.extend('dataModels.JSON', {
+
+        this.pipeline.forEach(function(sources, pipe) {
+            var dataSource = sources[pipe.name];
+
+            if (dataSource) {
+                if (dataSource.sorts) {
+                    dataSource.set(this.getSortedColumnIndexes().slice());
+                }
+
+                if (dataSource.apply) {
+                    dataSource.apply(options);
+                }
+            }
+        }.bind(this, this.sources));
+
+        reselectGridRowsBackedBySelectedDataRows.call(this);
+    },
+
+    /**
+     * @summary
+     * @desc
      *
-     * Branches are created when a pipe specifies a name in `parent`.
      * @param {object[]} [dataSource] - Array of uniform objects containing the grid data. Passed as 1st param to constructor of first data source object in the pipeline. If omitted, the previous data source will be re-used.
      * @param {string[]} [dataFields] - Array of field names. Passed as 2nd param to constructor of first data source object in the pipeline. If omitted (along with `dataSource`), the previous fields array will be re-used.
      * @param {string[]} [dataCalculators] - Array of field names. Passed as 3rd param to constructor of first data source object in the pipeline. If omitted (along with `dataSource`), the previous calculators array will be re-used.
      * @memberOf dataModels.JSON.prototype
      */
     setData: function(dataSource, dataFields, dataCalculators) {
-        this.resetSources();
+        this.source.setData(dataSource, dataFields, dataCalculators);
+        this.applyAnalytics();
+        //requeue ??
+    },
 
-        if (!dataSource) {
-            var source = this.source;
-            if (!source) {
-                throw 'Expected dataSource.';
-            }
-            dataSource = source.data;
-            dataFields = source.fields;
-            dataCalculators = source.calculators;
-        }
+    /**
+     * @summary Instantiates the data source pipeline.
+     * @desc Each new pipe is created using the supplied constructor and a reference to the previous data source in the pipeline. A reference to each new pipe is added to `this` dataModel as a property using the pipe's `name`.
+     *
+     * The last pipe is assigned the synonym `this.dataSource`.
+     *
+     * Branches are created when a pipe specifies a name in `parent`.
+     * @param {boolean} [stopApply] - Whether or not to run applyAnalytics
+     * @memberOf dataModels.JSON.prototype
+     */
+    setPipeline: function(stopApply) {
+        var dataSource = this.source,
+            mainPipeTip;
 
-        this.pipeline.forEach(function(sources, pipe, index) {
+        this.pipeline.forEach(function(sources, pipe) {
             var DataSource = analytics[pipe.type];
 
             pipe.name = pipe.name || getDataSourceName(pipe.type);
 
-            if (index === 0 && pipe.name !== 'source') {
-                throw 'Expected pipeline to begin with source.';
-            }
-
             if (pipe.parent) {
```

Comment:
We really don't need to support `parent` any more. This was just an accommodation to trying to have one immutable pipeline that served both the needs of aggreate data and regular data. Now that we have a constructed pipeline, we can get rid of this!